### ISSUE TITLE
Adding support for Indexes when using gen-sqlddl

### ISSUE THREAD

**Issue by Dariusms** [linkml/linkml#1910]
Currently no indexes are generated for columns. This can be a performance problem for mid and large size data sets.

At a bare minimum it would be really useful to add indexes to foreign key columns that are created by default since those are frequently the joins. However, more importantly it would be useful to be able to specify which slots/attributes in the data modeled should be indexed. Something like marking them as being heavily used, or sort, or searchable, etc.

**Comment by sierra-moxon** [comment]
@Dariusms - which database platform are you running the DDL on?  (I agree index DDL generation would be great! -- as an aside postgres does auto-generate indexes on any column with a primary key designation).

**Comment by Dariusms** [comment]
Thanks for the reply, we primarily use Oracle.

**Comment by cmungall** [comment]
It sounds like we should explicitly add an index (at the sqla level). sqla
should be smart enough not to turn these into actual indexes where the dbms
already does this. See
https://stackoverflow.com/questions/8890738/sqlalchemy-does-column-with-foreignkey-creates-index-automatically

**Comment by Dariusms** [comment]
Yes, that would address the foreign key scenario, I would like to also add the ability to specify that a particular slot or attribute should be indexed. From the user data model developer perspective it might be something like marking them as being either heavily used, or used for sort, or commonly used for searches, etc. 

Alternatively, we can take the approach that all slots and attributes could be indexed by default. But then we need to provide a mechanism for marking an attribute or slot as not being indexed, perhaps a slot property indicating it is strictly descriptive in nature.

**Comment by cmungall** [comment]
A few complementary approaches:

1. I think we need a non-unique analog of:

   https://linkml.io/linkml-model/latest/docs/unique_keys

While a non-unique analog would not have any semantics in terms of
constraints, it could still serve as a signifier that there is a particular
tuple of slots that are semantically meaningful in some way, and have these
be indexed by default.

This would take some time to percolate, it would need to go in 1.8.x or
later

2. Fine grained control at the annotation level

e.g.

classes:
  MyClass
    attributes:
      s1:
        ...
        annotations:
          index: true

This wouldn't require any metamodel changes, annotations would serve as
hints that generators are free to ignore

3. Fine grained control at the CLI level

`gen-sqlddl --index MyClass.s1 --index ...`

PRs for 2 or 3 would be acceptable in the short term. We can still pursue 1.

**Comment by Dariusms** [comment]
Thank you for that feedback. It was very informative. Yes, 1 is very interesting. Instead of the terminology of non-unique, may I propose something along the line of important_slots. But I do like it, it is quite clean and nice. Option 2 is the one I expected to see. It does not strike me as being as bad because I think there are some other annotations that I can conceive of that are missing that could fit under there like sizing (#1915).

**Comment by cmungall** [comment]
OK good, we can proceed with 2.

Although we can put whatever we like under annotations, it is good to have a separate schema for this. This is one of the more advanced features of linkml that you can schematize extensions like this. See https://linkml.io/linkml/schemas/annotations.html#annotation-validation

Would you like to start sketching out the idea of what kinds of annotation keys you'd want to use and what their expected values would be? You can just sketch this out here loosely if you like (but things will go faster if you can come up with an actual schema for it :-)

**Comment by Dariusms** [comment]
Thanks for the information on annotation validation, just to clarify, approach #1 is the long term / preferred approach, with the annotation approach being  more of a short term placeholder? If so, for the long term solution I would recommend:
1) Frequently_used:
       - name1
       - ...
2) Frequently_sorted:
      - name1
      - ...
3) Frequently_filtered:
      - name1
      - ...
For the annotations, as per your request I will try to put together a schema. Roughly speaking, my thoughts were split between easy implementation vs minimize the rework with the short term solution. As such it could be as simple as:
1) annotations:
      index: values, ascending, descending

or it could be more complicated like so:
2) annotations
      Frequently_sorted:
        - name1
      Frequently_filtered:
        - name1

**Comment by VictoriaSavageNNL** [comment]
Resurrecting this issue: We will be adding a field at the individual slot level in accordance with the annotation style mentioned above. To enable multi-column indexes, we will add a field within the schema class definition, I can expand on this
